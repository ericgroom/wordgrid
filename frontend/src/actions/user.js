export const SET_TOKEN = "SET_TOKEN";
export const SENT_AUTH = "SENT_AUTH";
export const REQUEST_SET_NICKNAME = "REQUEST_SET_NICKNAME";
export const SET_NICKNAME = "SET_NICKNAME";
export const SET_USERID = "SET_USERID";

export const setNickname = nickname => ({
  type: REQUEST_SET_NICKNAME,
  nickname
});

export const setUserId = id => ({
  type: SET_USERID,
  id
});
