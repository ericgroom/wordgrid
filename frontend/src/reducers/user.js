import {
  SET_NICKNAME,
  REQUEST_SET_NICKNAME,
  SET_USERID,
  CONFIRM_AUTH
} from "../actions";
const initialState = {
  nickname: null,
  nicknameRequested: false,
  userId: null,
  authConfirmed: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_NICKNAME:
      return { ...state, nickname: action.nickname, nicknameRequested: false };
    case REQUEST_SET_NICKNAME:
      return { ...state, nicknameRequested: true };
    case SET_USERID:
      return { ...state, userId: action.id };
    case CONFIRM_AUTH:
      return { ...state, authConfirmed: true };
    default:
      return state;
  }
};
