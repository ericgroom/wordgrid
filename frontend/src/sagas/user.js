import jwt_decode from "jwt-decode";
import { eventChannel } from "redux-saga";
import { call, all, take } from "redux-saga/effects";
import { putFrom, awaitAuthIfNeeded } from "./index";
import {
  SENT_AUTH,
  SET_NICKNAME,
  REQUEST_SET_NICKNAME,
  SET_TOKEN,
  setUserId,
  confirmAuth,
  CONFIRM_AUTH
} from "../actions";

function* userFlow(socket) {
  const authChannel = yield call(authSocketChannel, socket);
  const userChannel = yield call(userSocketChannel, socket);
  yield all([
    putFrom(userChannel),
    putFrom(authChannel),
    userActionListener(socket)
  ]);
}

function authSocketChannel(socket) {
  return eventChannel(emitter => {
    if (localStorage.authToken) {
      socket.emit("auth", localStorage.authToken, success => {
        emitter(confirmAuth(success));
        const { userId } = jwt_decode(localStorage.authToken);
        emitter(setUserId(userId));
      });
      emitter({ type: SENT_AUTH, new: false });
    } else {
      socket.emit("new auth", success => {
        emitter(confirmAuth(success));
      });
      emitter({ type: SENT_AUTH, new: true });
    }
    return () => {};
  });
}

function userSocketChannel(socket) {
  return eventChannel(emit => {
    socket.on("token", token => {
      localStorage.authToken = token;
      emit({ type: SET_TOKEN });
    });
    socket.on("nickname", nickname => {
      emit({ type: SET_NICKNAME, nickname });
    });
    return () => {
      console.log("in unsub");
    };
  });
}

function* userActionListener(socket) {
  yield call(awaitAuthIfNeeded);
  while (true) {
    console.log("userActionListener waiting...");
    const action = yield take(REQUEST_SET_NICKNAME, CONFIRM_AUTH);
    console.log("userActionListener taking action", action.type);
    switch (action.type) {
      case REQUEST_SET_NICKNAME:
        console.log("setting!");
        socket.emit("nickname", action.nickname);
        break;
      case CONFIRM_AUTH:
        console.log(`confirmed auth. success? ${action.success}`);
        break;
      default:
        throw new Error(
          `userActionListener saga took action but has no handler for: ${
            action.type
          }`
        );
    }
  }
}

export default userFlow;
