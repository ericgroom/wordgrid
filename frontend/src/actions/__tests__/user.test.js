import { setNickname, setUserId, confirmAuth } from "../user";

describe("user action creators", () => {
  it("setNickname", () => {
    expect(setNickname("Eric")).toMatchSnapshot();
  });
  it("setUserId", () => {
    expect(setUserId(123)).toMatchSnapshot();
  });
  it("confirmAuth", () => {
    expect(confirmAuth(true)).toMatchSnapshot();
  });
});
