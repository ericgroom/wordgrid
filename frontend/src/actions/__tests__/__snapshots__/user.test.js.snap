// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user action creators confirmAuth 1`] = `
Object {
  "success": true,
  "type": "CONFIRM_AUTH",
}
`;

exports[`user action creators setNickname 1`] = `
Object {
  "nickname": "Eric",
  "type": "REQUEST_SET_NICKNAME",
}
`;

exports[`user action creators setUserId 1`] = `
Object {
  "id": 123,
  "type": "SET_USERID",
}
`;
